File: receptor/algoritmos/crc32.js
// Algoritmo CRC-32 para deteccion de errores.
// Autor: Fernando Hernandez (23645)
//
// Mismo esquema que emisor/algoritmos/crc32.py: polinomio estandar de CRC-32
// (IEEE 802.3) 0x04C11DB7, registro inicial 0xFFFFFFFF y complemento final,
// procesando la trama bit a bit (MSB primero). El CRC de 32 bits viene
// concatenado al final del mensaje binario; n_datos permite retirar el padding
// agregado a mensajes de 32 bits o menos.

const NOMBRE = "crc32";
const POLINOMIO = 0x04c11db7;
const BITS_CRC = 32;

function crc(bits) {
  let registro = 0xffffffff;
  for (const bit of bits) {
    const msb = (registro >>> (BITS_CRC - 1)) & 1;
    registro = (registro << 1) >>> 0; // fuerza aritmetica sin signo de 32 bits
    if (msb ^ Number(bit)) registro = (registro ^ POLINOMIO) >>> 0;
  }
  return (registro ^ 0xffffffff) >>> 0;
}

function conPadding(bitsDatos) {
  if (bitsDatos.length > BITS_CRC) return bitsDatos;
  return bitsDatos.padEnd(BITS_CRC + 1, "0");
}

function calcularIntegridad(bitsDatos) {
  if (!bitsDatos) throw new Error("El mensaje binario esta vacio");
  const datos = conPadding(bitsDatos);
  return datos + crc(datos).toString(2).padStart(BITS_CRC, "0");
}

// Recalcula el CRC y lo compara con el recibido. CRC-32 detecta, no corrige:
// estado es 'ok' o 'error'. `nDatos` es la longitud original del mensaje.
function verificarIntegridad(trama, nDatos) {
  if (trama.length <= BITS_CRC) {
    throw new Error("La trama debe tener mas de 32 bits");
  }
  const datos = trama.slice(0, -BITS_CRC);
  const crcRecibido = trama.slice(-BITS_CRC);
  const crcCalculado = crc(datos).toString(2).padStart(BITS_CRC, "0");
  if (crcCalculado !== crcRecibido) {
    return { estado: "error", bitsDatos: "" };
  }
  return {
    estado: "ok",
    bitsDatos: nDatos != null ? datos.slice(0, nDatos) : datos,
  };
}

module.exports = {
  NOMBRE,
  BITS_CRC,
  calcularIntegridad,
  verificarIntegridad,
};

File: emisor/algoritmos/crc32.py
"""Algoritmo CRC-32 para deteccion de errores.

Autor: Fernando Hernandez (23645)

Usa el polinomio estandar de 32 bits de CRC-32 (IEEE 802.3):
0x04C11DB7. El registro inicia en 0xFFFFFFFF y el residuo final se
complementa, procesando la trama bit a bit (MSB primero). El CRC de 32 bits se
concatena al final del mensaje binario original.

La trama debe tener n > 32 bits; si el mensaje es menor se agregan 0s de
padding al final antes de calcular el CRC. Para retirar el padding en el
receptor se utiliza n_datos (longitud original del mensaje).
"""

NOMBRE = "crc32"
POLINOMIO = 0x04C11DB7
BITS_CRC = 32
MASCARA = 0xFFFFFFFF


def _crc(bits: str) -> int:
    """Calcula el CRC-32 de una cadena de bits, bit a bit (MSB primero)."""
    registro = MASCARA  # valor inicial: todos 1s
    for bit in bits:
        msb = (registro >> (BITS_CRC - 1)) & 1
        registro = (registro << 1) & MASCARA
        if msb ^ int(bit):
            registro ^= POLINOMIO
    return registro ^ MASCARA  # complemento final


def _con_padding(bits_datos: str) -> str:
    """Agrega 0s al final si el mensaje tiene 32 bits o menos (n > 32)."""
    if len(bits_datos) > BITS_CRC:
        return bits_datos
    return bits_datos + "0" * (BITS_CRC + 1 - len(bits_datos))


def calcular_integridad(bits_datos: str) -> str:
    """Concatena el CRC-32 (32 bits) al final del mensaje binario."""
    if not bits_datos:
        raise ValueError("El mensaje binario esta vacio")
    datos = _con_padding(bits_datos)
    return datos + format(_crc(datos), f"0{BITS_CRC}b")


def verificar_integridad(trama: str, n_datos=None) -> dict:
    """Recalcula el CRC del lado del receptor y lo compara con el recibido.

    Devuelve un dict con:
      estado: 'ok' | 'error' (CRC-32 detecta, no corrige)
      bits_datos: mensaje original sin padding ni CRC (si estado es 'ok')
    `n_datos` es la longitud original del mensaje, para retirar el padding.
    """
    if len(trama) <= BITS_CRC:
        raise ValueError("La trama debe tener mas de 32 bits")
    datos, crc_recibido = trama[:-BITS_CRC], trama[-BITS_CRC:]
    crc_calculado = format(_crc(datos), f"0{BITS_CRC}b")
    if crc_calculado != crc_recibido:
        return {"estado": "error", "bits_datos": ""}
    if n_datos is not None:
        datos = datos[:n_datos]
    return {"estado": "ok", "bits_datos": datos}
